src/devices/machine/pxa255.cpp

```diff
@@ -539,116 +766,93 @@ TIMER_CALLBACK_MEMBER(pxa255_periphs_device::ostimer_match_tick)
-			m_ostimer_regs.osmr[0] = data;
-			if (m_ostimer_regs.oier & PXA255_OIER_E0)
-			{
-				m_ostimer_regs.timer[0]->adjust(attotime::from_hz(3846400) * (m_ostimer_regs.osmr[0] - m_ostimer_regs.oscr), 0);
-			}
-			break;
-		case PXA255_OSMR1:
-			LOGMASKED(LOG_OSTIMER, "pxa255_ostimer_w: OS Timer Match Register 1: %08x & %08x\n", data, mem_mask);
-			m_ostimer_regs.osmr[1] = data;
-			if (m_ostimer_regs.oier & PXA255_OIER_E1)
-			{
-				m_ostimer_regs.timer[1]->adjust(attotime::from_hz(3846400) * (m_ostimer_regs.osmr[1] - m_ostimer_regs.oscr), 1);
-			}
-			break;
-		case PXA255_OSMR2:
-			LOGMASKED(LOG_OSTIMER, "pxa255_ostimer_w: OS Timer Match Register 2: %08x & %08x\n", data, mem_mask);
-			m_ostimer_regs.osmr[2] = data;
-			if (m_ostimer_regs.oier & PXA255_OIER_E2)
-			{
-				m_ostimer_regs.timer[2]->adjust(attotime::from_hz(3846400) * (m_ostimer_regs.osmr[2] - m_ostimer_regs.oscr), 2);
-			}
-			break;
-		case PXA255_OSMR3:
-			LOGMASKED(LOG_OSTIMER, "pxa255_ostimer_w: OS Timer Match Register 3: %08x & %08x\n", data, mem_mask);
-			m_ostimer_regs.osmr[3] = data;
-			if (m_ostimer_regs.oier & PXA255_OIER_E3)
-			{
-				//m_ostimer_regs.timer[3]->adjust(attotime::from_hz(3846400) * (m_ostimer_regs.osmr[3] - m_ostimer_regs.oscr), 3);
-			}
-			break;
-		case PXA255_OSCR:
-			LOGMASKED(LOG_OSTIMER, "pxa255_ostimer_w: OS Timer Count Register: %08x & %08x\n", data, mem_mask);
-			m_ostimer_regs.oscr = data;
-			break;
-		case PXA255_OSSR:
-			LOGMASKED(LOG_OSTIMER, "pxa255_ostimer_w: OS Timer Status Register: %08x & %08x\n", data, mem_mask);
-			m_ostimer_regs.ossr &= ~data;
-			ostimer_irq_check();
-			break;
-		case PXA255_OWER:
-			LOGMASKED(LOG_OSTIMER, "pxa255_ostimer_w: OS Timer Watchdog Enable Register: %08x & %08x\n", data, mem_mask);
-			m_ostimer_regs.ower = data & 0x00000001;
-			break;
-		case PXA255_OIER:
-		{
-			LOGMASKED(LOG_OSTIMER, "pxa255_ostimer_w: OS Timer Interrupt Enable Register: %08x & %08x\n", data, mem_mask);
-			m_ostimer_regs.oier = data & 0x0000000f;
-			for (int index = 0; index < 4; index++)
-			{
-				if (m_ostimer_regs.oier & (1 << index))
-				{
-					//m_ostimer_regs.timer[index]->adjust(attotime::from_hz(200000000) * m_ostimer_regs.osmr[index], index);
-				}
-			}
-			break;
-		}
-		default:
-			LOGMASKED(LOG_OSTIMER | LOG_UNKNOWN, "pxa255_ostimer_w: Unknown address: %08x = %08x & %08x\n", PXA255_OSTMR_BASE_ADDR | (offset << 2), data, mem_mask);
-			break;
+		m_ostimer_regs.timer[Which]->adjust(attotime::from_hz(3846400) * (m_ostimer_regs.osmr[Which] - m_ostimer_regs.oscr), Which);
 	}
 }
 
-/*
-
-  PXA255 Interrupt registers
+u32 pxa255_periphs_device::tmr_oscr_r(offs_t offset, u32 mem_mask)
+{
+	const u32 data = m_ostimer_regs.oscr;
+	LOGMASKED(LOG_OSTIMER, "%s: tmr_oscr_r: OS Timer Count Register: %08x & %08x\n", machine().describe_context(), data, mem_mask);
+	// free-running 3.something MHz counter.  this is a complete hack.
+	m_ostimer_regs.oscr += 0x300;
+	return m_ostimer_regs.oscr;
+}
```

Comment:
This one needs a `machine().side_effects_disabled()` check to avoid issues with debugger memory views.

Reply:
I've fixed this in a more elegant way by bringing over the more accurate OS-timer bookkeeping code from sa1110.cpp.